instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
         """Call :meth:`full_clean` before saving."""
         self.full_clean()
         super(ValidateModelMixin, self).save(*args, **kwargs)
+
+
+class ClassProperty(property):
+    """ Same as built-in 'property' global but also works when accessed as a class attribute """
+    def __get__(self, cls, owner):
+        return self.fget.__get__(None, owner)()  # pylint: disable=no-member
+
+
+class ResourceState:
+    """
+    A [finite state machine] state class representing the overall state of a resource,
+    such as a server, a database, a DNS entry, etc.
+    """
+    # state_id: A string uniquely identifying this state
+    state_id = None
+
+    @ClassProperty
+    @classmethod
+    def name(cls):
+        """
+        name: Human-readable name of this state, suitable for display as "Status: _______"
+
+        Doesn't need to be a property; override in the subclass using just 'name = "Name"'
+        """
+        return cls.__name__
+
+    @ClassProperty
+    @classmethod
+    def description(cls):
+        """
+        description: Human-readable explanation of this state (1-2 sentences)
+
+        Defaults to the state class's docstring, but you can override it.
+        Doesn't need to be a property; override in the subclass using just 'description = "..."'
+        """
+        return cls.__doc__.strip()
+
+    # Constants:
+    PROGRESS_UNKNOWN = -1
+
+    def __str__(self):
+        return "{} [{}]".format(self.name, self.state_id)
+
+    def __eq__(self, obj):
```

Comment:
It's generally recommended to implement `__hash__()` as well when implementing `__eq__()` (since having objects that compare equal return different hashes breaks assumptions).  Simply return the hash of `type(self)` should do the trick.